src/System.Drawing.Common/src/System/Drawing/Icon.Unix.cs

```diff
@@ -709,77 +709,163 @@ private void InitFromStreamWithSize(Stream stream, int width, int height)
-                // 256x256 icons are decoded as 0x0 (width and height are encoded as BYTE)
-                // and we ignore them just like MS does (at least up to fx 2.0) 
-                // Added: storing data so it can be saved back
-                if ((ide.width == 0) && (ide.height == 0))
-                    ide.ignore = true;
-                else
-                    ide.ignore = false;
+                            iidata.iconColors = new uint[numColors];
+                            for (int j = 0; j < numColors; j++)
+                                iidata.iconColors[j] = bihReader.ReadUInt32();
 
-                iconDir.idEntries[i] = ide;
+                            try
+                            {
+                                // XOR mask is immediately after ColorTable and its size is
+                                // icon height* no. of bytes per line
+
+                                // icon height is half of BITMAPINFOHEADER.biHeight, since it contains
+                                // both XOR as well as AND mask bytes
+                                int iconHeight = bih.biHeight / 2;
+
+                                // bytes per line should should be uint aligned
+                                int numBytesPerLine = checked((((bih.biWidth * bih.biPlanes * bih.biBitCount) + 31) >> 5) << 2);
+
+                                // Determine the XOR array Size
+                                int xorSize = checked(numBytesPerLine * iconHeight);
+                                iidata.iconXOR = new byte[xorSize];
+                                int nread = bihReader.Read(iidata.iconXOR, 0, xorSize);
+                                if (nread != xorSize)
+                                {
+                                    string msg = string.Format("{0} data length expected {1}, read {2}", "XOR", xorSize, nread);
+                                    throw new ArgumentException(msg, nameof(stream));
+                                }
+
+                                // Determine the AND array size
+                                numBytesPerLine = checked((((bih.biWidth) + 31) & ~31) >> 3);
+                                int andSize = checked(numBytesPerLine * iconHeight);
+                                iidata.iconAND = new byte[andSize];
+                                nread = bihReader.Read(iidata.iconAND, 0, andSize);
+                                if (nread != andSize)
+                                {
+                                    string msg = string.Format("{0} data length expected {1}, read {2}", "AND", andSize, nread);
+                                    throw new ArgumentException(msg, nameof(stream));
+                                }
+                            }
+                            catch (OverflowException)
+                            {
+                                throw new ArgumentException(nameof(stream));
```

Comment:
If you make the OverflowException nested, it will include the line on which the overflow occurred? Or you could just let the OverflowException propagate out (the argument name isn't that important I guess).

Reply:
I guess having argument exception with the SOE as an inner exception would be enough right? To have the same behavior between win and Unix?

Reply:
You mean OE? Nested works

Reply:
Yes I meant OE.